app/src/ui/create-branch/create-branch-dialog.tsx

```diff
@@ -259,14 +308,18 @@ export class CreateBranch extends React.Component<
     this.updateBranchName(name)
   }
 
-  private updateBranchName(branchName: string) {
+  private async updateBranchName(branchName: string) {
     const alreadyExists =
       this.props.allBranches.findIndex(b => b.name === branchName) > -1
 
     const currentError = alreadyExists
-      ? new Error(`A branch named ${branchName} already exists`)
+      ? { error: new Error(`A branch named ${branchName} already exists`), isBlocking: true }
       : null
 
+    if (!currentError) {
+      await this.checkBranchRules(branchName)
```

Comment:
Now that we've introduced an async operation here we would typically begin by issuing a state update to reflect the branch name change and then after the async operation check to make sure it hasn't changed while we were out fetching data from the API so that we don't end up in a race condition where the slowest API response "wins"

Same applies within the checkBranchRules method

Reply:
I think I get what you're saying to do, I added that in https://github.com/desktop/desktop/pull/16904/commits/edf4ba0a91234a66ca86deba69ff217e0a88b2c2. Is that all I need? I think this covers all of the async edge cases now.

Reply:
Actually, that commit just doesn't work, because `if (branchName === this.state.branchName)` is always false. I guess I'm not sure how you're saying to change it then. I could leave in the other change from https://github.com/desktop/desktop/commit/edf4ba0a91234a66ca86deba69ff217e0a88b2c2 but remove that specific check since it's always false?

Reply:
After discussion with some of the team, I'm going to try pulling in a package to perform the fnmatch-style matching and update to do these checks client-side, which would make this no longer async. That will come in a future PR though. We'll merge this one in as-is, then my next PR will either render this point moot or fix the logic.